src/main/java/fr/raluy/chocoratage/KeyBuffer.java

```diff
@@ -37,49 +36,99 @@ public KeyBuffer(boolean strict) {
+     * @param keyChar
+     * @return true if the buffer changed or false as the character was neither a letter nor a digit and couldn't be appended
+     */
+    public boolean append(char keyChar) {
+        boolean result;
+        if (result = isLetterOrDigit(keyChar)) {
+            currentWord.append(Character.toLowerCase(keyChar));
 
+            int currentSize = currentWord.length();
             if (currentSize > MAX_WORD_SIZE) {
                 currentWord.delete(0, currentSize - MAX_WORD_SIZE); //cut the buffer back to MAX_SIZE chars
             }
-        } else if (isBackspace(keyEvent)) {
-            if (isCurrentWordExists()) {
-                currentWord.delete(currentSize - 1, currentSize);
-            }
-        } else if (isCurrentWordExists()) { // let's start a new word
+        }
+        return result;
+    }
+
+    /**
+     * Removes the trailing char of the buffer.
+     * of course this way of working won't work correctly if someone types in 2 separators then backspaces twice => one character too many is removed
+     * @return true if the buffer changed
+     */
+    public boolean backspace() {
+        if (isCurrentWordExists()) {
+            int currentSize = currentWord.length();
+            currentWord.delete(currentSize - 1, currentSize);
+        }
+        else if(!previousWords.isEmpty()) { // here currentWord has already a zero-length
+            String previousWord = previousWords.removeLast();
+            currentWord.append(previousWord, 0, previousWord.length()-1);
+        } else {
+            return false;
+        }
+        return true;
+    }
+
+    public void newWord() {
+        if (isCurrentWordExists()) {
             if (previousWords.size() == MAX_WORDS) {
                 previousWords.removeFirst();
             }
             previousWords.add(currentWord.toString());
             currentWord.setLength(0);
         }
-        // TODO handle paste the best we can
     }
 
-    private boolean isBackspace(NativeKeyEvent keyEvent) {
-        return keyEvent.getKeyCode() == NativeKeyEvent.VC_BACKSPACE;
+    /**
+     * @param keyChar
+     * @return
+     */
+    public static boolean isLetterOrDigit(char keyChar) {
+        return Character.isLetterOrDigit(keyChar);
     }
 
-    private boolean isLetterOrDigit(NativeKeyEvent keyEvent) {
-        if (keyEvent.isActionKey()) {
-            return false;
-        }
+    public static boolean isBackspace(char keyChar) {
+        return keyChar == KeyEvent.VK_BACK_SPACE; // NOT NativeKeyEvent.VC_BACKSLASH, because we are dealing with a key char here (\b == 8), NOT a key code (0x000E)
+    }
 
-        String keyText = NativeKeyEvent.getKeyText(keyEvent.getKeyCode());
+    /**
+     * http://www.physics.udel.edu/~watson/scen103/ascii.html
+     *
+     * @param keyChar
+     * @return true if the input char is in the range [0x00-0x1F] or DEL (0x7F) or space (0x20)
+     */
+    public static boolean isControlCharOrSpace(char keyChar) {
```

Comment:
Is this safe to assume? Aren't we using UTF8 and not ascii?

Reply:
Yeah but they are ligned up to 0x7F :)
Read the 2nd § of [](https://en.wikipedia.org/wiki/UTF-8)